File: Lab6/src/main/java/com/example/lab6/CarRepository.java
package com.example.lab6;

import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class CarRepository {
    private List<Car> carList;
    private static final String filename = "cars.h2";

    public CarRepository() {
        this.carList = new ArrayList<>();
        loadCars();
    }

    public List<Car> getCarList() {
        return carList;
    }

    public void addCar(Car car) {
        carList.add(car);
        saveCars();
    }

    public void removeCar(Car car) {
        carList.remove(car);
        saveCars();
    }

    public List<Car> getAvailableCars() {
        return carList.stream()
                .filter(car -> !car.getIsRented())
                .collect(Collectors.toList());
    }

    public List<Car> getRentedCars() {
        return carList.stream()
                .filter(Car::getIsRented)
                .collect(Collectors.toList());
    }

    public List<Car> getCarsFreeInMonth(int month) {
        return carList.stream()
                .filter(car -> car.getIsRented() && car.getRentalDate().plusMonths(car.getRentalPeriod()).getMonthValue() == month)
                .collect(Collectors.toList());
    }

    public List<Car> getFilteredCars(String brand, String licensePlate) {
        return carList.stream()
                .filter(car -> (brand.isEmpty() || car.getBrand().equals(brand)) && (licensePlate.isEmpty() || car.getLicensePlate().equals(licensePlate)))
                .collect(Collectors.toList());
    }


    private void saveCars() {
        try {
            FileOutputStream fos = new FileOutputStream(filename);
            ObjectOutputStream oos = new ObjectOutputStream(fos);
            oos.writeObject(carList);
            oos.close();
            fos.close();
        } catch (IOException ioe) {
            ioe.printStackTrace();
        }
    }

    private void loadCars() {
        try {
            FileInputStream fis = new FileInputStream(filename);
            ObjectInputStream ois = new ObjectInputStream(fis);
            carList = (ArrayList<Car>) ois.readObject();
            ois.close();
            fis.close();
        } catch (FileNotFoundException fnfe) {
            System.out.println("No previous cars found. Initializing an empty list.");
            carList = new ArrayList<>();
        } catch (IOException ioe) {
            ioe.printStackTrace();
        } catch (ClassNotFoundException c) {
            System.out.println("Class not found");
            c.printStackTrace();
        }
    }
}

File: Lab6_kotlin/Lab6/src/main/kotlin/com/example/lab6/CarRepository.kt
package com.example.lab6

import java.io.*
import java.util.stream.Collectors

class CarRepository {
    val carList: MutableList<Car> = mutableListOf()

    init {
        loadCars()
    }

    fun addCar(car: Car) {
        carList.add(car)
        saveCars()
    }

    fun removeCar(car: Car) {
        carList.remove(car)
        saveCars()
    }

    val availableCars: List<Car>
        get() = carList.stream()
                .filter { car: Car -> !car.isRented }
                .collect(Collectors.toList())
    val rentedCars: List<Car>
        get() = carList.stream()
                .filter(Car::isRented)
                .collect(Collectors.toList())

    fun getCarsFreeInMonth(month: Int): List<Car> {
        return carList.stream()
                .filter { car: Car -> car.isRented && car.rentalDate!!.plusMonths(car.rentalPeriod.toLong()).monthValue == month }
                .collect(Collectors.toList())
    }

    fun getFilteredCars(brand: String, licensePlate: String): List<Car> {
        return carList.stream()
                .filter { car: Car -> (brand.isEmpty() || car.brand == brand) && (licensePlate.isEmpty() || car.licensePlate == licensePlate) }
                .collect(Collectors.toList())
    }

    private fun saveCars() {
        try {
            val fos = FileOutputStream(filename)
            val oos = ObjectOutputStream(fos)
            oos.writeObject(carList)
            oos.close()
            fos.close()
        } catch (ioe: IOException) {
            ioe.printStackTrace()
        }
    }

    private fun loadCars() {
        try {
            val fis = FileInputStream(filename)
            val ois = ObjectInputStream(fis)
            val loadedCars = ois.readObject() as ArrayList<Car>
            ois.close()
            fis.close()

            carList.clear()
            carList.addAll(loadedCars)
        } catch (fnfe: FileNotFoundException) {
            println("No previous cars found. Initializing an empty list.")
            carList.clear()
        } catch (ioe: IOException) {
            ioe.printStackTrace()
        } catch (c: ClassNotFoundException) {
            println("Class not found")
            c.printStackTrace()
        }
    }

    companion object {
        private const val filename = "cars.h2"
    }
}
